### PR TITLE
`Plugin.isDevDependency` if exclusively in `dev_dependencies`

### PR DESCRIPTION
**Pull request by matanlurey**
Work towards https://github.com/flutter/flutter/issues/56591.

I explicitly want an LGTM from @andrewkolos @jmagman @jonahwilliams before merging.

---

After this PR, `<Plugin>.isDevDependency` is resolved based on the following logic, IFF:

- The plugin comes from a package _A_ listed in the app's package's `dev_dependencies: ...`
- The package _A_ is not a normal dependency of any transitive non-dev dependency of the app

See [`compute_dev_dependencies_test.dart`](https://github.com/flutter/flutter/blob/51676093a3b67434405aacf4d56088573863630a/packages/flutter_tools/test/general.shard/compute_dev_dependencies_test.dart) for probably the best specification of this behavior.

We (still) do not write the property to disk (i.e. it never makes it to `.flutter-plugins-dependencies`), so there is no impact to build artifacts at this time; that would come in a follow-up PR (and then follow-up follow-up PRs for the various build systems in both Gradle and Xcode to actually use that value to omit dependencies).

Some tests had to be updated; for the most part it was updating the default `ProcessManager` because a call to `dart pub deps --json` is now made in code that computes what plugins are available, but there should be no change in behavior.

_/cc @jonasfj @sigurdm for FYI only (we talked on an internal thread about this; see https://github.com/dart-lang/sdk/issues/56968)._

_/cc @camsim99 @cbracken @johnmccutchan for visibility on the change._

### CODE REVIEW

**Comment by matanlurey** [comment]
Shoot, there are still `commands.shard` tests that need updates, I'll work on those.

**Review by jonahwilliams — APPROVED** [review]
So i think the overall approach SGTM. I don't really have much to nit about the code, but I don't really work in the tool anymore.

LGTM

**Comment by matanlurey** [comment]
Updated tests in `commands.shard`.

I'm slowed down by running into test failures that only occur locally but seem unrelated:

- https://github.com/flutter/flutter/issues/157627
- https://github.com/flutter/flutter/issues/157629

File: packages/flutter_tools/lib/src/compute_dev_dependencies.dart
Comment (jmagman) [inline packages/flutter_tools/lib/src/compute_dev_dependencies.dart:157]: Is an infinite loop possible?

Comment (matanlurey) [reply packages/flutter_tools/lib/src/compute_dev_dependencies.dart:157]: No, clarified in both the comment and with a local variable above.

**Review by jmagman — APPROVED** [review]

**Comment by matanlurey** [comment]
Sigh, this is blocked by https://github.com/flutter/flutter/issues/102983 and https://github.com/flutter/flutter/issues/73870.

**Comment by matanlurey** [comment]
Will need to be resolved by adding a flag (`--implicit-pubspec-resolution`) and make this code rely on it:

https://github.com/flutter/flutter/issues/157819

Blocked/moving to draft for now.

**Comment by matanlurey** [comment]
Please re-review, I merged this with https://github.com/flutter/flutter/pull/157879.

**Review by andrewkolos — DISMISSED** [review]
There's a lot of unintentional-looking formatting here :/

**Comment by matanlurey** [comment]
> There's a lot of unintentional-looking formatting here :/

Yeah my mistake, I really messed up my default workspace - will fix that going forward.

**Comment by matanlurey** [comment]
Ping - looking for another once-over so I can unblock https://github.com/flutter/flutter/pull/158009.

Thanks!

**Review by jmagman — APPROVED** [review]
Sorry for the delay.

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Pull request merged**
